### PR TITLE
Use CPP instead of Switches

### LINKED ISSUE #360: Transition from switch pre-processing to CPP ifdefs
Instead of using the current paradigm of choosing switches, and then processing the *.ftn code to *.F90 code, we will convert all code to *.F90 and use the switches to define CPP if defs.  

This work is planned for June 2021.  To prepare for this, we recommend making pull requests with updates as soon as possible. In addition, removal of switches that are no longer needed have been proposed and will be cleaned out first.

### PR DESCRIPTION
**Pull request by JessicaMeixner-NOAA**
# Pull Request Summary
Use CPP ifdefs instead of switches for pre-processing and convert all code from ftn to F90 files. (ftn->src) 

## Description
This PR  provides a utility in tools/ that converts the WW3 switches into CPP ifdefs. The build is also updated for this change. The utility in model/tools/ftn2src.sh is run to convert the ftn to a src directory.  

Co-author: @ukmo-ccbunney 

### Issue(s) addressed
- fixes #360

### Check list  
* Is your feature branch up to date with the authoritative repository (NOAA/develop)? yes (+PR454)
* Make sure you have checked the [checklist for a developer submitting to develop](https://github.com/NOAA-EMC/WW3/wiki/Code-Management#checklist-for-a-developer-submitting-to-develop) and [updating version number](https://github.com/NOAA-EMC/WW3/wiki/Code-Management#checklist-for-updating-version-number) Version number is not currently updated for this. 

 
* Reviewers: @aliabdolali @ukmo-ccbunney @thesser1 @mickaelaccensi 

### Commit Message
* Use CPP ifdefs instead of switches for pre-processing and convert all code from ftn to F90 files. (ftn->src)  The tools for converting ftn->src can be found in the model/tools directory. 

### Testing
* How were these changes tested? matrix_ncep regression tests 
* Are the changes covered by regression tests? (If not, why? Do new tests need to be added?) yes 
* If a new feature was added, was a new regression test added? existing tests are sufficient 
* Have regression tests been run? yes
* Which compiler / HPC you used to run the regression tests in the PR? hera.intel 
* Please provide the summary output of matrix.comp (_matrix.Diff.txt_, _matrixCompFull.txt_ and _matrixCompSummary.txt_):   (Note these are compared to output from PR454 as its included in this PR)

[matrixCompFull.txt](https://github.com/NOAA-EMC/WW3/files/7178235/matrixCompFull.txt)
[matrixCompSummary.txt](https://github.com/NOAA-EMC/WW3/files/7178236/matrixCompSummary.txt)
[matrixDiff.txt](https://github.com/NOAA-EMC/WW3/files/7178237/matrixDiff.txt)

Please indicate the expected changes in the outputs ([excluding the known list of non-identical tests](https://github.com/NOAA-EMC/WW3/wiki/How-to-use-matrix.comp-to-compare-regtests-with-master#4-look-at-results)).

```
**********************************************************************
********************* non-identical cases ****************************
**********************************************************************
mww3_test_03/./work_PR3_UQ_MPI_e                     (1 files differ)
mww3_test_03/./work_PR2_UQ_MPI_d2                     (8 files differ)
mww3_test_03/./work_PR3_UNO_MPI_d2_c                     (9 files differ)
mww3_test_03/./work_PR3_UNO_MPI_d2                     (10 files differ)
mww3_test_03/./work_PR1_MPI_e                     (1 files differ)
mww3_test_03/./work_PR2_UNO_MPI_d2                     (8 files differ)
mww3_test_03/./work_PR1_MPI_d2                     (8 files differ)
mww3_test_03/./work_PR3_UQ_MPI_d2_c                     (8 files differ)
mww3_test_03/./work_PR3_UQ_MPI_d2                     (8 files differ)
mww3_test_03/./work_PR3_UQ_MPI_e_c                     (1 files differ)
mww3_test_07/./work_PR3_UQ                     (3 files differ)
ww3_tp2.10/./work_MPI_OMPH                     (6 files differ)
ww3_ufs1.1/./work_b                     (0 files differ)
ww3_ufs1.2/./work_b                     (0 files differ)
ww3_ufs1.3/./work_a                     (1 files differ)
```

* Please list which labels code managers should add to indicate code changes:    
There should be no output changes

### CODE REVIEW

**Comment by mickaelaccensi** [comment]
it's compiling well with both intel and gnu compilers for me.
I'm running the regtests with intel to check if everything goes well

**Comment by JessicaMeixner-NOAA** [comment]
> it's compiling well with both intel and gnu compilers for me.
> I'm running the regtests with intel to check if everything goes well

@mickaelaccensi thanks for letting me know! This does not include all the recent code changes but hopefully will not negatively affect the running of regtests too much.  Right now we need to resolve how to move forward on this discussion topic: https://github.com/NOAA-EMC/WW3/discussions/474 so we can resolve the issues @ukmo-ccbunney has had before this can be merged

**Comment by mickaelaccensi** [comment]
> it's compiling well with both intel and gnu compilers for me.
> I'm running the regtests with intel to check if everything goes well

regtests ran correctly

**Comment by JessicaMeixner-NOAA** [comment]
The update for w3init switch definitions has been updated in this branch (as well as in the PR #475).  There's also been a few more updates for cleaning up things and the code has been updated to be up to date with the develop branch.

**Comment by JessicaMeixner-NOAA** [comment]
Also, I did confirm that we are running the ftn->.F90 pre-processor 3 times so any double switch should be properly handled.

**Comment by JessicaMeixner-NOAA** [comment]
@thesser1 @mickaelaccensi @ukmo-ccbunney can you try to build and run at least 1 regression test again?  There has been an update to make ad3 a templated file which allows for a compiler specific processing and removes the need to define a fortran compiler during the set-up phase.  It's my hope that this will solve the issues that Chris had with earlier versions. 

@aliabdolali I have tested this with the ufs-weather-model regression tests on hera, orion, gaea and wcoss_dell (see: https://github.com/JessicaMeixner-NOAA/ufs-weather-model/tree/ww3_swtich2cpp) I'll open a corresponding PR there shortly assuming this fixes the issues Chris was having, and then we're just waiting for the last PR and any additional testing anyone wants to do is welcome.  I also ran the regression tests again on both hera and orion and got the expected results on hera.

**Comment by JessicaMeixner-NOAA** [comment]
@thesser1 @mickaelaccensi @ukmo-ccbunney any updates? Are things still working for you?  We are still waiting for PR 483 to go through first, but if there are any major roadblocks it'd be great to know about them sooner/later so I can make sure to get them fixed so we're ready to go when it's this PRs turn.

**Comment by mickaelaccensi** [comment]
matrix ran well again for me

**Comment by JessicaMeixner-NOAA** [comment]
> matrix ran well again for me

Thanks for testing again @mickaelaccensi !

**Comment by ukmo-ccbunney** [comment]
@JessicaMeixner-NOAA - I've run `ww3_tp1.1` successfully on both the CrayFortran and GNU compilers (using the Cray `ftn` wrapper script).

I did have to make one small adjustment to cmplr.env though for the CrayFortran compiler in the `ukmo_cray` section:

Lines 411/412:
```
-  cppad3flag2='#'
-  cppad3flag3=' '
+  cppad3flag2=' '
+  cppad3flag3='#'
```

File: model/bin/cmplr.env [hunk 1]
```diff
@@ -383,4 +406,11 @@ if [ "$cmplr" == "ukmo_cray" ] || [ "$cmplr" == "ukmo_cray_debug" ] || \
 
   # system-dependant options
   # N/A
+  #Flags for CPP: 
+  cppad3procflag='-eP'
+  cppad3flag2='#'
```
Comment (ukmo-ccbunney) [inline model/bin/cmplr.env]: For ukmo_cray section, this needs to be
```
cppad3flag2=' '
cppad3flag3='#'
```

**Review by ukmo-ccbunney — CHANGES_REQUESTED** [review]
@JessicaMeixner-NOAA - I've run `ww3_tp1.1` successfully on both the CrayFortran and GNU compilers (using the Cray `ftn` wrapper script).

I did have to make one small adjustment to lines 411/412 of cmplr.env though for the CrayFortran compiler in the `ukmo_cray` section (see review comment).

File: model/bin/cmplr.env (hunk 1 above)
Comment (JessicaMeixner-NOAA) [reply model/bin/cmplr.env]: Updated

Comment (JessicaMeixner-NOAA) [reply model/bin/cmplr.env]: Thanks for testing this!

**Comment by JessicaMeixner-NOAA** [comment]
Okay, the latest develop has been merged in.  This branch is ready for further testing.  

@aliabdolali since the ufs-weather-model results have changed, I cannot make a PR without changing results for this PR until your PR is complete.   I will make a subsequent ufs-weahter-model PR with this branch, but as I think this is a fairly large change it should be done separately.   I will make a ufs-weather-model issue for this.

**Comment by JessicaMeixner-NOAA** [comment]
All of the code here is up to date and is ready for everyone to do a full set of regression tests.  I'll start a new set here at EMC as well.   

@ukmo-ccbunney @thesser1 (or Mary) @mickaelaccensi  can you please all test again?   @aliabdolali this is ready for your review as well.  I'll post new matrix.comp & ufs-weather-model regtest updates asap.

**Comment by ukmo-ccbunney** [comment]
Full regression test matrix completed ok with only the usual suspects being non-B4B:

```
**********************************************************************
********************* non-identical cases ****************************
**********************************************************************
mww3_test_03/./work_PR3_UQ_MPI_d2                     (8 files differ)
mww3_test_03/./work_PR3_UQ_MPI_d2_c                     (9 files differ)
mww3_test_03/./work_PR3_UNO_MPI_d2                     (10 files differ)
mww3_test_03/./work_PR1_MPI_d2                     (7 files differ)
mww3_test_03/./work_PR2_UNO_MPI_d2                     (10 files differ)
mww3_test_03/./work_PR3_UNO_MPI_d2_c                     (8 files differ)
mww3_test_03/./work_PR2_UQ_MPI_d2                     (9 files differ)
mww3_test_03/./work_PR3_UNO_MPI_e                     (1 files differ)
mww3_test_03/./work_PR2_UNO_MPI_e                     (1 files differ)
```

**Comment by aliabdolali** [comment]
> Full regression test matrix completed ok with only the usual suspects being non-B4B:
> 
> ```
> **********************************************************************
> ********************* non-identical cases ****************************
> **********************************************************************
> mww3_test_03/./work_PR3_UQ_MPI_d2                     (8 files differ)
> mww3_test_03/./work_PR3_UQ_MPI_d2_c                     (9 files differ)
> mww3_test_03/./work_PR3_UNO_MPI_d2                     (10 files differ)
> mww3_test_03/./work_PR1_MPI_d2                     (7 files differ)
> mww3_test_03/./work_PR2_UNO_MPI_d2                     (10 files differ)
> mww3_test_03/./work_PR3_UNO_MPI_d2_c                     (8 files differ)
> mww3_test_03/./work_PR2_UQ_MPI_d2                     (9 files differ)
> mww3_test_03/./work_PR3_UNO_MPI_e                     (1 files differ)
> mww3_test_03/./work_PR2_UNO_MPI_e                     (1 files differ)
> ```

Thanks @ukmo-ccbunney 
I put a hold on new PRs before we test and merge this PR.

**Comment by JessicaMeixner-NOAA** [comment]
The regression tests on hera are still in the queue.  I'm working on getting the matrix.comp output for hera.  The ufs-weather-model tests have run on some but not all platforms I have access too.  Will post results when I have them.  Thanks @ukmo-ccbunney for running the tests at your end. Do you approve of this PR now? (It still says you request changes).

**Review by ukmo-ccbunney — APPROVED** [review]

**Comment by JessicaMeixner-NOAA** [comment]
Issue #494 is stopping me from completing these tests.  We will likely need some sort of bug-fix in develop before we can proceed with testing this.

**Comment by thesser1** [comment]
Regression results from my intel compiler
**********************************************************************
********************* non-identical cases ****************************
**********************************************************************
mww3_test_03/./work_PR2_UQ_MPI_d2                     (8 files differ)
mww3_test_03/./work_PR2_UNO_MPI_d2                     (10 files differ)
mww3_test_03/./work_PR3_UNO_MPI_d2_c                     (6 files differ)
mww3_test_03/./work_PR3_UQ_MPI_d2                     (9 files differ)
mww3_test_03/./work_PR3_UNO_MPI_d2                     (6 files differ)
mww3_test_03/./work_PR1_MPI_d2                     (9 files differ)
mww3_test_03/./work_PR3_UQ_MPI_d2_c                     (8 files differ)
mww3_test_07/./work_PR3_UQ                     (3 files differ)
ww3_ufs1.1/./work_d                     (1 files differ)
ww3_ufs1.1/./work_c                     (1 files differ)
ww3_ufs1.1/./work_c_npl                     (1 files differ
[matrixCompFull.txt](https://github.com/NOAA-EMC/WW3/files/7373791/matrixCompFull.txt)
[matrixCompSummary.txt](https://github.com/NOAA-EMC/WW3/files/7373792/matrixCompSummary.txt)
[matrixDiff.txt](https://github.com/NOAA-EMC/WW3/files/7373793/matrixDiff.txt)
)

**Comment by aliabdolali** [comment]
> Regression results from my intel compiler
> 
> ********************* non-identical cases ****************************
> 
> mww3_test_03/./work_PR2_UQ_MPI_d2 (8 files differ) mww3_test_03/./work_PR2_UNO_MPI_d2 (10 files differ) mww3_test_03/./work_PR3_UNO_MPI_d2_c (6 files differ) mww3_test_03/./work_PR3_UQ_MPI_d2 (9 files differ) mww3_test_03/./work_PR3_UNO_MPI_d2 (6 files differ) mww3_test_03/./work_PR1_MPI_d2 (9 files differ) mww3_test_03/./work_PR3_UQ_MPI_d2_c (8 files differ) mww3_test_07/./work_PR3_UQ (3 files differ) ww3_ufs1.1/./work_d (1 files differ) ww3_ufs1.1/./work_c (1 files differ) ww3_ufs1.1/./work_c_npl (1 files differ)

Thanks @thesser1 
I assume all tests passed on your hpc, right?
Could you upload the matrix.comp stdout files?

**Comment by JessicaMeixner-NOAA** [comment]
I'll post comparisons in the next hour --- so far all tests are looking good for me. Then we just need @mickaelaccensi to confirm things are working

**Comment by mickaelaccensi** [comment]
it's running. I'll tell you the results once it's done

**Comment by JessicaMeixner-NOAA** [comment]
Regtests passed on hera, except for known issues: 

```
**********************************************************************
********************* non-identical cases ****************************
**********************************************************************
mww3_test_03/./work_PR2_UQ_MPI_d2                     (8 files differ)
mww3_test_03/./work_PR3_UNO_MPI_d2_c                     (8 files differ)
mww3_test_03/./work_PR3_UNO_MPI_d2                     (10 files differ)
mww3_test_03/./work_PR2_UNO_MPI_d2                     (8 files differ)
mww3_test_03/./work_PR1_MPI_d2                     (13 files differ)
mww3_test_03/./work_PR3_UQ_MPI_d2_c                     (8 files differ)
mww3_test_03/./work_PR3_UQ_MPI_d2                     (8 files differ)
mww3_test_07/./work_PR3_UQ                     (3 files differ)
ww3_tp2.10/./work_MPI_OMPH                     (7 files differ)
ww3_tp2.16/./work_MPI_OMPH                     (5 files differ)
ww3_ufs1.1/./work_d                     (0 files differ)
ww3_ufs1.2/./work_b                     (0 files differ)
ww3_ufs1.3/./work_a                     (1 files differ)

```

Regression tests also ran on orion, with known issues of netcdf errors

[matrixDiff.txt](https://github.com/NOAA-EMC/WW3/files/7347168/matrixDiff.txt)
[matrixCompSummary.txt](https://github.com/NOAA-EMC/WW3/files/7347169/matrixCompSummary.txt)
[matrixCompFull.txt](https://github.com/NOAA-EMC/WW3/files/7347170/matrixCompFull.txt)



ufs-weather-model passed for gaea.intel, hera.intel and Orion.intel for wave tests except for known point output not matching.  The branch with tests is here: https://github.com/JessicaMeixner-NOAA/ufs-weather-model/tree/feature/ww3cpp

**Comment by mickaelaccensi** [comment]
all regtests passed except usual mww3_test_03
[matrixCompFull.txt](https://github.com/NOAA-EMC/WW3/files/7371651/matrixCompFull.txt)
[matrixCompSummary.txt](https://github.com/NOAA-EMC/WW3/files/7371652/matrixCompSummary.txt)
[matrixDiff.txt](https://github.com/NOAA-EMC/WW3/files/7371653/matrixDiff.txt)

**Review by mickaelaccensi — APPROVED** [review]

**Review by thesser1 — APPROVED** [review]

**Review by aliabdolali — APPROVED** [review]

**Pull request merged**